### ISSUE TITLE
lambdify fails with xarray dataarrays broadcasting in 1.1

### ISSUE THREAD

**Issue by gerritholl** [sympy/sympy#12940]
`sympy.lambidy(…)` fails with `ValueError: operands could not be broadcast together` in sympy 1.1 when I pass xarray DataArrays with a broadcasting intention.   In both cases I'm passing `modules=numpy` explicitly, so I'm not sure this is related to #11468.  The behaviour in sympy 1.0 is intended and shown below.

Is this due to an intended backward-incompatible change?  There is no difference if I pass `modules=[xarray, numpy]`.  How can I get back the broadcasting behaviour from `sympy 1.0`?

```
$ cat mwe.py
#!/usr/bin/env python3.6
import sympy
import xarray
import numpy

print("sympy", sympy.__version__)

A = xarray.DataArray([1, 2, 3], dims=("x",))
B = xarray.DataArray([4, 5], dims=("y",))

(a, b) = sympy.symbols("a b")
f = sympy.lambdify((a, b), "a*b", dummify=False, modules=numpy)
print(f(A, B))
$ pip install sympy==1.0
Collecting sympy==1.0
  Using cached sympy-1.0.tar.gz
Requirement already satisfied: mpmath>=0.19 in /dev/shm/gerrit/venv/stable-3.6/lib/python3.6/site-packages (from sympy==1.0)
Installing collected packages: sympy
  Found existing installation: sympy 1.1
    Uninstalling sympy-1.1:
      Successfully uninstalled sympy-1.1
  Running setup.py install for sympy ... done
Successfully installed sympy-1.0
$ ./mwe.py
sympy 1.0
<xarray.DataArray (x: 3, y: 2)>
array([[ 4,  5],
       [ 8, 10],
       [12, 15]])
Dimensions without coordinates: x, y
15:27:51:9931:gholl@host292:~$ pip install sympy==1.1
Collecting sympy==1.1
  Using cached sympy-1.1.tar.gz
Requirement already satisfied: mpmath>=0.19 in /dev/shm/gerrit/venv/stable-3.6/lib/python3.6/site-packages (from sympy==1.1)
Installing collected packages: sympy
  Found existing installation: sympy 1.0
    Uninstalling sympy-1.0:
      Successfully uninstalled sympy-1.0
  Running setup.py install for sympy ... done
Successfully installed sympy-1.1
$ ./mwe.py
sympy 1.1
Traceback (most recent call last):
  File "./mwe.py", line 13, in <module>
    print(f(A, B))
  File "/dev/shm/gerrit/venv/stable-3.6/lib/python3.6/site-packages/sympy/utilities/lambdify.py", line 441, in wrapper
    return funcarg(*[namespace['asarray'](i) for i in argsx], **kwargsx)
  File "<string>", line 1, in <lambda>
ValueError: operands could not be broadcast together with shapes (3,) (2,)
```

**Comment by gerritholl** [comment]
This appears to be a consequence of the `asarray` call.  Can I disable this?  There is no equivalent in `xarray`, I just want to put the `xarray` datasets in unaltered.

**Comment by bjodah** [comment]
Thanks for mwe.py, running `git bisect` with it:
```
$ git bisect start master sympy-1.0
$ git bisect run python3 mwe.py
```
finds e63dbdf4ccbbb9e457eef3bf7db73d4e8641b1e2

**Comment by bjodah** [comment]
I tried changing `asarray` to `asanyarray` but it didn't help. Any thought @richardotis ?

**Comment by richardotis** [comment]
Is it an acceptable workaround to pass `A.values` and `B.values` as arguments instead of the original DataArrays?

**Comment by asmeurer** [comment]
Maybe only call asarray for int/float inputs.

**Comment by gerritholl** [comment]
Not really.  Passing `A.values` or `B.values` is not enough; I'd need something like `A.values[:, numpy.newaxis]` and `B.values[numpy.newaxis, :]` to make it work, but the one of the main reasons I'm using `xarray` at all is the way it broadcasts named dimensions and considers coordinates.  And although it's rather easy to workaround in this minimum working example, my actual case is not at all minimal so doing this across the board would be very involved.

The `asarray` line complicates the use of any custom types.  For those inheriting from `numpy.ndarray` that would be solved by `asanyarray` (for example, for masked arrays), but for others it isn't.  One other example I can think of is [`pint` arrays with units](http://pint.readthedocs.io/en/0.7.2/), but anyone with a custom type that implements `__add__` is going to have trouble passing their types to `lambdify`.

**Comment by asmeurer** [comment]
Or check if it's an `ndarray` first and only call if it isn't? What is the correct way to do that?

**Comment by richardotis** [comment]
@gerritholl We will keep working this issue, but to try to help you with your specific case (since the release cycle for sympy can be long), would this workaround work?
`f(*[x.values for x in xarray.align(*args)])`

**Comment by asmeurer** [comment]
We are planning on a 1.1.1 bugfix release soon. I've made my proposed fix #12946 against the `1.1.1` branch.

**Comment by gerritholl** [comment]
@richardotis Not with `xarray.align`, as my case is more about dimensions than indices.  If I instead use [`xarray.broadcast`](http://xarray.pydata.org/en/stable/generated/xarray.broadcast.html) the MWE no longer fails, but to make the workaround complete I'll need one more step to turn the resulting `ndarray` back into a `DataArray` with the correct dimension names, coordinates, attributes etc.  I think I can get there.
